### DISCUSSION TITLE
ArgumentOutOfRangeException: x Parameter name: Negative value passed to Sqrt

### DISCUSSION THREAD

**Discussion by TalhaEjaz03** [bepu/bepuphysics1#18] [Q&A]
Hi Norbo.
We are getting this strange fatal error of the Fix64 class.
This is produced when the colliders are small and a lot.
Also when they are smashing together.

Error Details are below

ArgumentOutOfRangeException: x
Parameter name: Negative value passed to Sqrt
FixMath.NET.Fix64.Sqrt (FixMath.NET.Fix64 x) (at Assets/!BEPU/bepuphysics1int-master/FixedMath.Net/src/Fix64.cs:609)
BEPUphysics.Constraints.Collision.SlidingFrictionTwoAxis.SolveIteration () (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Constraints/Collision/SlidingFrictionTwoAxis.cs:180)
BEPUphysics.Constraints.SolverGroups.SolverGroup.SolveUpdateable (BEPUphysics.Constraints.SolverUpdateable item, System.Int32& activeConstraints) (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Constraints/SolverGroups/SolverGroup.cs:129)
BEPUphysics.Constraints.Collision.ConvexContactManifoldConstraint.SolveIteration () (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Constraints/Collision/ConvexContactManifoldConstraint.cs:207)
BEPUphysics.Constraints.SolverGroups.SolverGroup.SolveUpdateable (BEPUphysics.Constraints.SolverUpdateable item, System.Int32& activeConstraints) (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Constraints/SolverGroups/SolverGroup.cs:129)
BEPUphysics.Constraints.SolverGroups.SolverGroup.SolveIteration () (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Constraints/SolverGroups/SolverGroup.cs:163)
BEPUphysics.Constraints.Solver.UnsafeSolveIteration (BEPUphysics.Constraints.SolverUpdateable updateable) (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Constraints/Solver.cs:296)
BEPUphysics.Constraints.Solver.UpdateSingleThreaded () (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Constraints/Solver.cs:259)
BEPUphysics.MultithreadedProcessingStage.Update () (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/MultithreadedProcessingStage.cs:86)
BEPUphysics.Space.DoTimeStep () (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Space.cs:512)
BEPUphysics.Space.Update () (at Assets/!BEPU/bepuphysics1int-master/BEPUphysics/Space.cs:530)
PhysicsManager.Update () (at Assets/!BEPU/_BEPU_Scripts/BepuScps/PhysicsManager.cs:76)

[CustomCompoundBody.txt](https://github.com/bepu/bepuphysics1/files/5982116/CustomCompoundBody.txt)
[PhysicsManager.txt](https://github.com/bepu/bepuphysics1/files/5982117/PhysicsManager.txt)
[SphereAndBoxColl.txt](https://github.com/bepu/bepuphysics1/files/5982118/SphereAndBoxColl.txt)

**Comment by RossNordby** [comment]
That shouldn't be possible in the sliding friction solve step- the only square root taken is guarded to be positive. It's possible that something odd is happening in the fixed point math and producing a NaN-equivalent, which is then interpreted as negative for the purposes of that error (don't know if that's true- I'm not familiar with FixedMath.Net).

I'd recommend either trying to reproduce it in the normal demos, or examining the state of the constraint at the exception and trace invalid state backwards until you find the cause.

**Reply by TalhaEjaz03** [reply]
As always, thanks for quick and elaborate reply!

Actually the problem persists whenever the objects collide and pass into each others bounding boxes. In our case (multiple compound bodies enclosed in a dynamic box). When the box is moving at a certain velocity the objects enclosed in, they start penetrating each other and cause them to throw the above exception(neg value passed to sqrt).

Whenever a complex compound body with normal to high velocity comes into contact with another compound body, they penetrate and cause errors. If by any method we could mitigate the passing through(penetration) it'll be a life saver!!!

THANKS A BUNCH NORDBY!

**Reply by RossNordby** [reply]
Enabling continuous collision detection and/or taking shorter and more frequent timesteps would be the best way to reduce penetration, but penetration shouldn't be causing that issue to begin with. I haven't observed something similar in the main branch, so this might be a quirk in the fixed point implementation.
